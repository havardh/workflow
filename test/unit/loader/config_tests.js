// @flow
/* eslint-env jest */
/* eslint-disable no-unused-expressions, import/no-extraneous-dependencies */
import os from 'os';
import { sandbox } from 'sinon';

import RequireWrapper from '../../../util/requireCompiled';
import load from '../../../loader/config';

const sinon = sandbox.create();

function ModuleNotFound() {
  this.code = 'MODULE_NOT_FOUND';
}

function FileReadFailed() {
  this.code = 'ENOENT';
}

describe('load(configFile)', () => {
  const expectedFile = { file: 'content' };

  const path = __dirname.replace('/test/unit', '');
  beforeEach(() => {
    sinon.stub(RequireWrapper, 'require');
    sinon.stub(os, 'homedir');
    sinon.stub(console, 'log');

    // $FlowTodo
    RequireWrapper.require.throws(new ModuleNotFound());

    os.homedir.returns('/home/user');
  });

  afterEach(() => {
    sinon.restore();
  });

  it('should throw error when all loaders fails', () => {
    const fn = () => load('file.js');
    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should throw loading error when reading all files fails', () => {
    // $FlowTodo
    RequireWrapper.require.throws(new FileReadFailed());

    const fn = () => load('file.js');
    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should rethrow error when loaded module fails', () => {
    // $FlowTodo
    RequireWrapper.require.throws(new Error('Module error'));

    const fn = () => load('file.js');

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should require absolute path', () => {
    RequireWrapper.require
      // $FlowTodo
      .withArgs('/path/to/file.js')
      .returns({ default: expectedFile });

    const file = load('/path/to/file.js');

    expect(file).toEqual(expectedFile);
  });

  it('should append missing .js to absolute path', () => {
    RequireWrapper.require
      // $FlowTodo
      .withArgs('/path/to/file.js')
      .returns({ default: expectedFile });

    const file = load('/path/to/file');

    expect(file).toEqual(expectedFile);
  });

  it('should require file from ~/.workflow', () => {
    RequireWrapper.require
      // $FlowTodo
      .withArgs('/home/user/.workflow/file.js')
      .returns({ default: expectedFile });

    const file = load('file.js');

    expect(file).toEqual(expectedFile);
  });

  it('should append missing .js when require from ~/.workflow', () => {
    RequireWrapper.require
      // $FlowTodo
      .withArgs('/home/user/.workflow/file.js')
      .returns({ default: expectedFile });

    const file = load('file');

    expect(file).toEqual(expectedFile);
  });

  it('should require file from workflow examples', () => {
    RequireWrapper.require
      // $FlowTodo
      .withArgs(`${path}/../examples/file.js`)
      .returns({ default: expectedFile });

    const file = load('file.js');

    expect(file).toEqual(expectedFile);
  });

  it('should append missing .js when require from workflow examples', () => {
    RequireWrapper.require
      // $FlowTodo
      .withArgs(`${path}/../examples/file.js`)
      .returns({ default: expectedFile });

    const file = load('file.js');

    expect(file).toEqual(expectedFile);
  });

  it('should require in specific order', () => {
    RequireWrapper.require
      // $FlowTodo
      .withArgs(`${path}/../examples/file.js`)
      .returns({ default: expectedFile });

    load('file.js');

    // $FlowTodo
    const calls = RequireWrapper.require.getCalls();

    expect(calls[0].args[0]).toEqual('file.js');
    expect(calls[1].args[0]).toEqual('/home/user/.workflow/file.js');
    expect(calls[2].args[0]).toEqual(`${path}/../examples/file.js`);
  });
});
