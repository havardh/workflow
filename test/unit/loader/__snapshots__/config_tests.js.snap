// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load(configFile) should rethrow error when loaded module fails 1`] = `"Module error"`;

exports[`load(configFile) should throw error when all loaders fails 1`] = `"Could not load file.js"`;
